### DISCUSSION TITLE
How do I access RDS database in the Reference Architecture from terraform machine?

### DISCUSSION THREAD

**Discussion by yorinasub17** [gruntwork-io/knowledge-base#398] [Help] [answered]
I'm trying to use the `postgresql` provider to provision a couple of database users onto every database that we deploy with the `rds` module from the service catalog.  It has to establish a client connection to the database to do this, and it's failing to connect to the RDS instances with a timeout.  Can anybody give me pointers on what I'll need to modify in our Gruntworks architecture (security groups?  network ACLs?  both?) to enable this?

---

<ins datetime="2022-04-28T14:33:44Z">
  <p><a href="https://gruntwork.zendesk.com/agent/tickets/108522">Tracked in ticket #108522</a></p>
</ins>

**Comment by yorinasub17** [comment, marked as answer]
The following are necessary to allow `terraform` access. Some of these can be skipped depending on your operating model (e.g., if you only use ECS Deploy Runner), but you will most likely want to enable all of them.

- Implement VPC peering between the `mgmt` VPC (where ECS Deploy Runner runs) and the `app` VPC, and link the private subnets of the `mgmt` VPC to the app.
    - This can be done using the peering parameters in the [vpc module](https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/modules/networking/vpc).
- Update the `allow_connections_from_cidr_blocks` input variable in the `rds` module to allow connections from private subnets of the `mgmt` VPC.
- If you are using OpenVPN to connect from your local machine, update the `allow_connections_from_cidr_blocks` to include the OpenVPN subnet. See https://github.com/gruntwork-io/knowledge-base/discussions/359 for more details on which CIDR to use.

**Reply by 0x2f0x00** [reply]
Thank you for the prompt response!

**Reply by 0x2f0x00** [reply]
> Update the allow_connections_from_cidr_blocks input variable in the rds module to allow connections from private subnets of the mgmt VPC.
@yorinasub17 does the deploy runner run from within the mgmt VPC then?

**Reply by yorinasub17** [reply]
Yup that is correct.

**Reply by 0x2f0x00** [reply]
> Implement VPC peering between the mgmt VPC (where ECS Deploy Runner runs) and the app VPC
@yorinasub17 why wouldn't I just make the routes straight to the persistence VPC? (since RDS is what I need to be able to reach)

**Reply by yorinasub17** [reply]
The RDS networking inherits the network of the subnets that you are deploying it in, and is not a separate entity in the cloud. So you can't create a straight network route cross VPC without a peering connection, or putting the RDS in the public network space.

**Reply by 0x2f0x00** [reply]
Got it thanks.  

One more quick question, I see the option for DNS setting up a DNS forwarder in the the module, I'm wondering if I need to enable that?  (As things stand now, non-public RDS endpoints resolve just fine from the app VPC of course.). 

If I do need to enable the DNS forwarder, will it matter which VPC (mgmt or app) is the "origin_vpc_id"?  If so, which should it be?

Thanks again, appreciate your knowledge on this.

**Reply by yorinasub17** [reply]
The DNS forwarder should not be necessary, as the RDS endpoint should resolve correctly across the peer. The forwarder is necessary if you have a private route53 hosted zone that is being used to manually manage the endpoint DNS records.

**Reply by 0x2f0x00** [reply]
Thank you.

This is more of a general Terra[grunt,form] question:

In my module that's deploying the RDS instance, I've got a block declaring `dependency "vpc"` for which the config path is the local `../../../networking/vpc`, and of course I have access to `dependency.vpc.outputs` to get hold of the needed `vpc_id`.  I can do the same to grab hold the `vpc-mgmt` details to get this peering connection in place.

The question I have is, can I use a `dependency` block to monkey patch the vpc that it configured?  Since this RDS module is the only thing in the account requiring the peering connection between app and mgmt I'd prefer to isolate the knowledge of that peering connection it to the RDS module itself, making it easier to read and understand in the future, encapsulating everything required.

Thanks again @yorinasub17 for any help you can provide.

**Reply by yorinasub17** [reply]
You can potentially do that by extending the `rds` module to add the VPC peering resources into the module, but at that point you would be [creating your own service catalog](https://docs.gruntwork.io/reference/services/intro/create-your-own-service-catalog). But it would not be possible using our existing service modules.
